development-guidelines/guidelines.md

```diff
@@ -3,113 +3,113 @@
-- **Direct implementation usage**: Consider configuring the implementation’s state variables with values that prevent it from being used directly, such as by setting a flag during construction that disables the implementation and causes all methods to revert. This is particularly important if the implementation also performs delegatecall operations because this opens up the possibility of unintended self destruction of the implementation.
-- **Immutable and constant variables**: These are embedded into the bytecode and can therefore get out of sync between the proxy and implementation. If the implementation has an incorrect immutable variable, this value may still be used even if the same variables are correctly set in the proxy’s bytecode.
-- **Contract Existence Checks**: All [low-level calls](https://docs.soliditylang.org/en/latest/control-structures.html?highlight=existence#error-handling-assert-require-revert-and-exceptions), not just delegatecall, will return true against an address with empty bytecode. This might cause callers to be misled to think that a call performed a meaningful operation when it did not or it might result in important safety checks being silently skipped. Be aware that while a contract’s constructor is running, its bytecode remains empty until the end of constructor execution. We recommend that you rigorously verify that all low-level calls are properly protected against nonexistent contracts, keeping in mind that most proxy libraries (such as the one written by Openzeppelin) do not perform contract existence checks automatically.
+- **Storage layout**: Proxy and implementation storage layouts must be the same. Instead of defining the same state variables for each contract, both should inherit all state variables from a shared base contract.
+- **Inheritance**: Be aware that the order of inheritance affects the final storage layout. For example, `contract A is B, C` and `contract A is C, B` will not have the same storage layout if both B and C define state variables.
+- **Initialization**: Immediately initialize the implementation. Implement an uninitialized implementation contract to prevent well-known disasters (and near disasters). A factory pattern can help ensure correct deployment and initialization and reduce front-running risks.
+- **Function shadowing**: Ensure the proxy’s function is called if the same method is defined on the proxy and implementation. Be mindful of `setOwner` and other administration functions commonly found on proxies.
+- **Direct implementation usage**: Configure implementation state variables with values that prevent direct use, such as setting a flag during construction that disables the implementation and causes all methods to revert. This is particularly important if the implementation also carries delegatecall operations, as this may lead to accidental self-destruction.
```

Comment:
```suggestion
- **Direct implementation usage**: Configure implementation state variables with values that prevent direct use, such as setting a flag during construction that disables the implementation and causes all methods to revert. This is particularly important if the implementation also performs delegatecall operations, as this may lead to unintended self-destruction of the implementation.
```